zfs-utils based zfs graph driver

Mic92 opened the pull request:
- Use go-zfs for robust commandline handling
- implements complete driver interface including the differ interface
- should also works on non-linux platforms

current status: works on my machine(tm)
I am looking for testers with different zfs setups to find more bugs

**Usage** (a little bit more verbose for testers):
- the following lines can be tested without messing up your current docker setup
- replace zroot/docker with your preferred zfs dataset

```
$ git clone https://github.com/Mic92/docker
$ cd docker
$ git checkout features/zfs-rebased
$ export AUTO_GOPATH=1
$ hack/make.sh dynbinary
root> zfs create -o com.sun:auto-snapshot=false -o dedup=off -o mountpoint=/var/lib/docker-zfs zroot/docker
root> bundles/1.4.1-dev/dynbinary/docker -p /var/run/docker-zfs.pid -H tcp://127.0.0.1:2376 -s zfs -g /var/lib/docker-zfs -d -D
# spawn another shell and go to your favorite docker project and hit
$ export DOCKER_HOST=tcp://127.0.0.1:2376
$ bundles/1.4.1-dev/dynbinary/docker --pidfile=/var/run/docker-zfs.pid build .
```

You can also overwrite the used dataset with `--storage-opt zfs.fsname=zroot/docker`, when starting the docker daemon.

Related to docker/docker#7901

PKGBUILD for archlinux: https://aur.archlinux.org/packages/docker-zfs/

**ZFS Quickstart** (from a loopback device)
1. Install zfs:
on Ubuntu:

```
# apt-add-repository --yes ppa:zfs-native/stable
# apt-get update
# apt-get install debootstrap spl-dkms zfs-dkms ubuntu-zfs
```

on Arch:
install `zfs-git` from `AUR`
1. Create your zpool:

```
# truncate -s 2G /tmp/zfs.img
# zpool create dockertest /tmp/zfs.img
```

Mic92 committed:
Implement Docker on ZFS

Signed-off-by: Arthur Gautier <baloo@gandi.net>
Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

bakins commented:
I hadn't been following this PR, but I'm on the team that writes/maintains go-zfs. Generally, as long as we don't break core functionality, we are open to PR's and/or help with needed features.

crosbymichael commented:
Assigned to @unclejack 

Just make sure we have build tags for the driver and check with @tianon on the deps.

LGTM

tianon commented on Dockerfile:
```diff
@@ -26,6 +26,10 @@
 MAINTAINER	Tianon Gravi <admwiggin@gmail.com> (@tianon)
 
+RUN	gpg --keyserver pgp.mit.edu --recv-keys "F6B0FC61"
+RUN	gpg --armor --export "F6B0FC61" | apt-key add -
```
This should instead just be:
```Dockerfile
RUN apt-key adv --keyserver pool.sks-keyservers.net --recv-keys E871F18B51E0147C77796AC81196BA81F6B0FC61
```

tianon commented on project/vendor.sh:
```diff
@@ -55,6 +55,8 @@ clone git github.com/docker/libtrust 230dfd18c232
 
 clone git github.com/Sirupsen/logrus v0.6.0
 
+clone git github.com/mistifyio/go-zfs f24c6bc9603ab5ad
```
Can we adjust this to use upstream's `v2.1.0` tag instead?  If that tag isn't recent enough, we should see if upstream's willing to tag a new one. :+1:

tianon commented:
Can we also add a little something to https://github.com/docker/docker/blob/master/project/PACKAGERS.md#optional-dependencies mentioning the lower bound on versions that are acceptable for the ZFS tools?

tiborvass commented:
@Mic92 Needs rebase + address @tianon's concerns, thanks!

Mic92 commented:
Changes
- rebased against master
- apply api changes in driver
- secure ppa gpg key import
- switch to go-zfs upstream release
- add packaging documentation

@tianon The driver does not use any new feature of zfs, as the minimum kernel required for docker is 3.8 and zfs-utils is bound to its kernel module version, there aren't any known incompatibilities. In theory 0.4.1 from 2009 would be the first usable zfsonlinux release, but one won't get it compiled with newer kernels.

thaJeztah commented:
I think it would be nice to have checks added to the [`contrib/check-config.sh`](https://github.com/docker/docker/blob/master/contrib/check-config.sh) script, so that users are able to check if all the requirements are in place to use ZFS.

Mic92 commented:
@thaJeztah done

thaJeztah commented:
@Mic92 that was fast, thanks!

tiborvass commented:
@unclejack @tianon @jfrazelle @LK4D4 please test and review :)

tianon commented on contrib/check-config.sh:
```diff
@@ -83,6 +83,14 @@ check_flags() {
+	else
+		wrap_bad "$1 command" 'missing'
+	fi
+}
```
I really like this! :+1:

I wonder if maybe we should have a `check_device` too (for consistency), something like:
```bash
check_device() {
    if [ -c "$1" ]; then
        wrap_good "$1" 'present'
    else
        wrap_bad "$1" 'missing'
    fi
}
```

(and then using it below for `/dev/zfs`, of course)

tianon commented on contrib/check-config.sh:
```diff
@@ -158,6 +166,16 @@ flags=(
+	fi
+	echo "  - $(check_command zfs)"
+	echo "  - $(check_command zpool)"
+}
```
If we do something like the `check_device` I mentioned above, this will be short enough that it could go back down in the block below and we wouldn't need this inconsistent `check_zfs` function. :smile:

tianon commented:
:+1: this is looking pretty solid from my PoV

Mic92 commented:
@tianon done

unclejack commented:
@Mic92 I'm going to test this right now and comment after review.

This will have to be squashed because there are a lot of small changes. The travis.yml removal commit doesn't need to be there either.

tianon commented on contrib/check-config.sh:
```diff
@@ -83,6 +83,22 @@ check_flags() {
+    wrap_good "$1" 'present'
+  else
+    wrap_bad "$1" 'missing'
+  fi
```
Tabs :wink:

tianon commented:
Other than indentation, `contrib` and `project` LGTM :+1: :heart:

unclejack commented:
I've seen multiple tests fail when running the tests with zfs as the graph driver:
https://gist.github.com/unclejack/80f77d72c3a87bb6fbf4

Mic92 commented:
@unclejack Thanks for testing, I have the classical "works-on-my-machine"-problem and need some more background informations to reproduce your failing tests:

could you give me:

* the output of the following commands:

```
$ zfs list
$ zpool list
$ uname -a
$ modinfo zfs | grep version
$ mount
$ zfs get all
```
* the command you used for testings, if it differs from `make test`
* the working directory of your docker daemon if it differ from `/var/lib/docker`
* command line args to docker, if you used `--storage-opt`

unclejack commented:
@Mic92 `make test` only tests with the `vfs` graph driver by default in docker-in-docker. It can't guess what graph drivers it should use to test when using Docker-in-Docker because the graph driver used on the host can be incompatible with the graph driver used by the Daemon running in the container.

`make test` will only tell you whether using the zfs graph driver on the host breaks the vfs driver in Docker-in-Docker.

In order to actually test with zfs, you need to modify the Makefile to bind mount /var/lib/docker-zfs in the container under /var/lib/docker (e.g.: `-v /var/lib/docker-zfs:/var/lib/docker`) and run `make test DOCKER_GRAPHDRIVER=zfs`.

The distribution details are:
Distro & version: Ubuntu 14.04
kernel: 3.18.5 generic from http://kernel.ubuntu.com/~kernel-ppa/mainline/
zfs: the zfs packages from https://launchpad.net/~zfs-native/+archive/ubuntu/stable added using the ppa (0.6.3, the version which was there yesterday)

zpool: single disk with one fs set up using the instructions you've provided
storage-opt: I haven't passed any arguments

I've confirmed zfs was used during the tests by running `zfs list` to see snapshots and file systems - they were showing up and they were being deleted as the tests were running.

Should you really need the exact versions and output for everything, I can provide these later.

You should be able to reproduce the errors I've seen by passing DOCKER_GRAPHDRIVER=zfs and bind mounting /var/lib/docker-zfs into the container.

Mic92 commented:
@unclejack thanks for clarification, I assumed it picks the best driver it found, just like it does, if it runs from commandline. I also did only focused on the unit tests I wrote.
Now I can reproduce the errors.

Mic92 commented:
Wouldn't it be helpful, if the jenkins instance also tests each graph driver?

unclejack commented:
@Mic92 You can also drop by the Freenode #docker-dev channel if you want to discuss. Hanging around that channel certainly wouldn't hurt.

unclejack commented:
@Mic92 That's being worked on by @jfrazelle. Testing these new graph drivers is only done after their code is merged.

jfrazelle commented:
@Mic92 it does test all graphdrivers, when we add zfs I will add a zfs node, but actually just a note its kinda really hard on our servers when you push to this branch w 32 commits, bc it tests all 32 commits, we can only run 9 at once...

tilgovi commented:
I just wanted to add that I'm playing around with this now and encountering no problems.

magnayn commented:
1.7.0 now?!

Boooo.

tilgovi commented:
Running fine here! Anything the community can do to help?

unclejack commented:
The last time this was tested with the full test suite, there were multiple failures when using this ZFS storage driver with Docker.

Not passing the full test suite means there are bugs which will show up in various circumstances and this could lead to data corruption. These bugs will then get reported one by one, instead of being fixed before the code is merged.

Those failed tests are a showstopper and that's the reason why this hasn't been merged already. A graph driver which doesn't pass the test suite when it's being used can't be merged..


If you want to help get this merged faster, please read the instructions above and send PRs to @Mic92's branch to help fix those problems faster.

tilgovi commented:
Totally understandable. Thanks for the explanation.

trisk commented:
You can use ```gopkg.in/mistifyio/go-zfs.v2``` as an import to avoid surprises.

Mic92 commented:
is gopkg.in a stable service?

tianon commented:
We vendor the whole dep, so that shouldn't be necessary.

Mic92 committed:
zfs: revert to NaiveGraphDriver for the moment

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

Mic92 commented:
@Novex Because I removed the native diff driver for the moment, your hardlink problem should be disappeared.

Novex commented:
This does indeed fix the issue! Is there any downside to using the naive graph driver instead of something zfs specific?

Mic92 commented:
@Novex The checks between container build steps are a bit slower. No difference in the runtime performance.

LK4D4 commented:
ping @Mic92 for rebase
@unclejack for test

Mic92 commented:
done

unclejack commented:
I'm currently testing this. My previous test environment with zfs was destroyed because of a kernel upgrade.

unclejack commented:
@Mic92 The tests are mostly passing now. The ones which are failing involve daemon restarts. The tests timed out for me after 30 minutes. This is most likely caused by poor disk performance.

I'm currently running the tests again.

Can you squash your commits, please?

tianon commented:
LGTM :heart:

unclejack commented:
@Mic92 The tests were mostly passing until the timeout was reached due to ZFS being slow on that disk. However, that VM isn't coming back up when the disk with the zpool used for testing is connected.
I'm not sure if this is a ZFS problem or if it's something related to one of the snapshots or ZFS file systems.

![zfs](https://cloud.githubusercontent.com/assets/1409668/6960179/e4cfe172-d928-11e4-8916-af04fa07741a.png)

Mic92 committed:
docs: move zfs.fsname option to storage option section

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

unclejack commented:
LGTM up to this point. There are serious performance issues because of ZFS on Linux, but we need to get this merged to let it evolve and become mature.

ping @duglin @icecrime @jfrazelle @LK4D4 @tiborvass @vbatts - for those of you who want to review this PR and have the time to do this, please take a look and comment

jfrazelle commented:
I will spin up a server and test and if its all good then it can be the one that tests via jenkins master :D builds after merge

unclejack commented:
@jfrazelle I'm sorry to disappoint, but getting this to the point where it's ready to run the tests all the time will take a bit more work. However, I believe getting there requires this code to be in Docker's source tree.

jfrazelle commented:
oooo i see sorry for misunderstanding

Mic92 committed:
zfs: add myself to MAINTAINERS

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+func (d *Driver) Create(id string, parent string) error {
+	mountPoint := path.Join(d.options.mountPath, "graph", id)
+	datasetName := d.ZfsPath(id)
+	dataset, err := zfs.GetDataset(datasetName)
```
what happens here when `err != nil`?

Mic92 replied:
if `dataset.Destroy(zfs.DestroyRecursiveClones)` returns an error, there are two options:
- if dataset was already destroyed (race condition between GetDataset and Destroy)
- it failed to destroying the dataset, in this case zfs.CreateFilesystem will also fail
I will use log.Warnf instead

calavera replied:
I'm talking about `GetDataset` returning an error. We're not checking that.

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+			"mountpoint": mountPoint,
+		})
+		return err
+	} else {
```
this `} else {` can be omitted:

```go
if parent == "" {
  _, err := zfs.CreateFilesystem(datasetName, map[string]string{
    "mountpoint": mountPoint,
  })
  return err
}
return cloneFilesystem(datasetName, d.ZfsPath(parent), mountPoint)
```

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+	dataset, err := zfs.GetDataset(d.ZfsPath(id))
+	if dataset == nil {
+		return "", err
+	} else {
```
this `} else {` can be omitted.

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+		dir := C.GoString(Cmnt.mnt_dir)
+		if err := syscall.Stat(dir, &stat); err != nil {
+			log.Debugf("[zfs] failed to stat '%s' while scanning for zfs mount: %v", dir, err)
+			continue // may fail on fuse file systems
```
Is it possible that this might end up creating an infinite loop on fuse? should we set a backoff strategy with a maximum number of attempts?

Mic92 replied:
I don't think so, it will just skip those file system. It is a foreach(entry: /proc/mounts) loop

calavera replied:
sounds good

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+func checkRootdirFs(rootdir string) error {
+	var buf syscall.Statfs_t
+	if err := syscall.Statfs(rootdir, &buf); err != nil {
+		fmt.Errorf("Failed to access '%s': %s", rootdir, err)
```
Should we return this error?

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+			{"error while getting pool", fmt.Sprintf("%v", err)},
+		}
+	}
+	var quota string
```
you can avoid the `} else {` initializing `quota := "no"` :smile:

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+	}
+	var quota string
+	if d.dataset.Quota == 0 {
+		quota = strconv.FormatUint(d.dataset.Quota, 10)
```
maybe:

```go
if d.dataset.Quota == 0 {
  quota = "0"
}
```

?

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+
+func cloneFilesystem(id, parent, mountpoint string) error {
+	parentDataset, err := zfs.GetDataset(parent)
+	if parentDataset == nil {
```
can you check `if err != nil` in this function's check? it's mostly for consistency.

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+
+func (d *Driver) Get(id, mountLabel string) (string, error) {
+	dataset, err := zfs.GetDataset(d.ZfsPath(id))
+	if dataset == nil {
```
can you please check `if err != nil`?

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	logger := Logger{}
+	zfs.SetLogger(&logger)
```
This two lines can be just one with `zfs.SetLogger(new(Logger))`

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	file, err := os.OpenFile("/dev/zfs", os.O_RDWR, 600)
+	defer file.Close()
```
this `defer` should go after the `if err != nil`

calavera replied:
It would also be nice if you checked the error returned by `file.Close()` instead of ignoring it:

https://golang.org/pkg/os/#File.Close

Mic92 replied:
According to close(2): "It is quite possible that errors on a previous write(2) operation are first reported at the final close()" and "...In particular close() should not be retried after an EINTR since this  may cause a reused descriptor from another thread to be close..." 

As there  is no write operation on this file descriptor, there is no benefit in checking for errors on this file descriptor.

btw.
grep -r "defer.*Close()" /usr/lib/go | grep -v "_test" | wc -l

calavera replied:
sounds good :+1:

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+			return C.GoString(Cmnt.mnt_fsname), nil
+		}
+	}
+	// should never happen
```
best final words :grimacing:

stevvooe replied:
adding: if you believe them, then panic.

Mic92 replied:
I think, the zfs mount could disappear between statfs and reading /proc/mounts or beween reading /proc/mounts and the Stats syscalls.
So returning an error should be ok here:
1. If the user has explicitly request the zfs driver, the deamon will be stopped with this error message
2. otherwise the next graphdriver will be tried.

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+func Init(base string, opt []string) (graphdriver.Driver, error) {
+	var err error
+	options, err := parseOptions(opt)
+	options.mountPath = base
```
this should go after the `if err != nil`

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	if _, err := exec.LookPath("zfs"); err != nil {
+		return nil, fmt.Errorf("zfs command is not available")
```
I think it should be better if you returned `nil, err` here instead of your own error.

unclejack replied:
I disagree, this is OK. I'd add err at the end of this error, not replace this error with err.

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+		case "zfs.fsname":
+			options.fsName = val
+		default:
+			return options, fmt.Errorf("Unknown option %s\n", key)
```
the `\n` at the end of the error is not necessary.

Mic92 committed:
zfs: refactor error handling

thanks to @calavera

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

calavera commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -0,0 +1,277 @@
+}
+
+var CprocMounts = C.CString("/proc/mounts")
+var CopenMod = C.CString("r")
```
Do these two variables need to be global and exported? I only see them being used inside `lookupZfsDataset` and I cannot find any other reference inside docker itself. I think they should be inside the function and freed properly after used.

calavera commented:
This is looking awesome, thanks @Mic92 for your infinite patience with my comments!

:zap: LGTM :zap:

LK4D4 commented on daemon/graphdriver/driver.go:
```diff
@@ -38,6 +38,7 @@ var (
 	priority = []string{
 		"aufs",
 		"btrfs",
+		"zfs",
```
Shouldn't it have lower priority due to experimental nature?

unclejack replied:
@LK4D4 Having it here is reasonable because:
1) devicemapper on a ZFS FS is a bad idea
2) ZFS is incompatible with overlay (most likely)
3) aufs still gets used on top of zfs if the kernel is aufs enabled

stevvooe commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -240,6 +240,11 @@ explains how to tune your existing setup without the use of options.
 
+The `zfs` driver is probably not fast as `btrfs` but has a longer track record
+on stability. Thanks to `Single Copy ARC` shared blocks between clones will be
+cached only ones. Use `docker -d -s zfs`. To select a different zfs filesystem
```
Spelling: "once".

Mic92 committed:
integration: add variable to set storage options for testing

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

Mic92 commented:
@jfrazelle I have added DOCKER_STORAGE_OPTS to make zfs driver a little bit more CI friendly.
I tried to create a temporary zpool for integration testing within docker on top on a sparse image, but it seems like zfs has a bug and does not see this image in the mount namespace of docker. So the idea is the following:

Before running the tests:

```
$ touch /tmp/zfs-image && truncate --size 521M /tmp/zfs-image
$ zpool create -f dockerzfstest /tmp/zfs-image
$ zfs set sync=disabled dockerzfstest # evil performance tuning
```

Starting the tests:

```
$ make DOCKER_GRAPHDRIVER=zfs DOCKER_STORAGE_OPTS="zfs.fsname=dockerzfstest"
```

After running the tests:

```
$ zpool destroy dockerzfstest
```

What I am currently working on, is not mounting zfs datasets by default. Because of the high number of mounts this seems to degrade the performance (the zfs tool has to read /proc/self/mounts everytime, which can takes several milliseconds according to sysdig - and probably affect other userland programs too).

jfrazelle commented:
Ah awesome thanks so much! After 1.6.0 is out I am going to start testing
this to see if we can get it in and help with anything that is a blocker :)

On Wed, Apr 15, 2015 at 12:44 AM, Jörg Thalheim <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> I have added
> DOCKER_STORAGE_OPTS to make zfs driver a little bit more CI friendly.
> I tried to create a temporary zpool for integration testing within docker
> on top on a sparse image, but it seems like zfs has a bug and does not see
> this image in the mount namespace of docker. So the idea is the following:
>
> Before running the tests:
>
> $ touch /tmp/zfs-image && truncate --size 521M /tmp/zfs-image
> $ zpool create -f dockerzfstest /tmp/zfs-image
> $ zfs set sync=disabled dockerzfstest # evil performance tuning
>
> Starting the tests:
>
> $ make DOCKER_GRAPHDRIVER=zfs DOCKER_STORAGE_OPTS="zfs.fsname=dockerzfstest"
>
> After running the tests:
>
> $ zpool destroy dockerzfstest
>
> What I am currently working on, is not mounting zfs datasets by default.
> Because of the high number of mounts this seems to degrade the performance
> (the zfs tool has to read /proc/self/mounts everytime, which can takes
> several milliseconds according to sysdig - and probably affect other
> userland programs too).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9411#issuecomment-93245057>.
>

Mic92 committed:
zfs: replace c for /proc/mounts parsing with go

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

Mic92 committed:
zfs: improve performance by using legacy mounts

instead of let zfs automaticly mount datasets, mount them on demand using mount(2).
This speed up this graph driver in 2 ways:
- less zfs processes needed to start a container
- /proc/mounts get smaller, so zfs userspace tools has less to read (which can
  a significant amount of data as the number of layer grows)

This ways it can be also ensured that the correct mountpoint is always used.

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "features/zfs-rebased" git@github.com:Mic92/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~8
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "features/zfs-rebased" git@github.com:Mic92/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~8
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

Mic92 commented:
@unclejack with latest commit I speed up the driver by using less zfs processes and mounting datasets on demand. The tests now take less than 30 minutes on my machine. However I would increase the timeout to at least 40 minutes for DOCKER_GRAPHDRIVER=zfs

Mic92 committed:
zfs: retrieve all filesystems on startup at once

The docker graph call driver.Exists() on initialisation for each filesystem in
the graph. This results will results in a lot `zfs get all` commands. To reduce
this, retrieve all descend filesystem at startup and cache it for later checks

Signed-off-by: Jörg Thalheim <joerg@higgsboson.tk>

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "features/zfs-rebased" git@github.com:Mic92/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~9
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

thaJeztah commented:
@Mic92 commit 6f89f90c8c077b5a364c8e634ab0e54431b6f88c isn't signed-off, so the tests won't be running (not sure it's important if they timed out anyway :))

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "features/zfs-rebased" git@github.com:Mic92/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~9
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

Mic92 commented:
@thaJeztah the current CI server uses GRAPH_DRIVER=vfs and will therefor not timeout.

maci0 commented:
+1

tiborvass commented:
@Mic92 sorry needs a rebase again :(

Mic92 commented:
@tiborvass rebased

tiborvass commented:
wow thanks for your reactiveness, really appreciated! Hope to get this in sooner rather than later!

jfrazelle commented:
oh man so sorry needss another rebase we are so close :)

Mic92 commented:
@jfrazelle done

tiborvass commented:
LGTM

tiborvass commented:
@Mic92 thank you so very much for your awesome work and even greater patience!

tiborvass commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -444,6 +449,17 @@ Currently supported options are:
     > daemon with a supported environment.
 
 ### Docker execdriver option
+Currently supported options of `zfs`:
```
@Mic92 of course i just realized that this should be in the storagedriver section, not execdriver....

Mic92 replied:
Sorry this has happened during rebasing.

tiborvass commented:
@moxiegirl @SvenDowideit @fredlf my personal apologies for merging without docs review. The author was very patient. Feel free to make corrections on this PR, I'll take care of them.

thaJeztah commented:
LOL, just noticed docs wasn't reviewed yet. Will have a look later and add notes

burke commented:
Exciting! One thought regarding possible optimization that's not possible on most other drivers:

If, immediately after cloning the snapshot to inherit the parent, you clone another snapshot of that brand-new dataset called "parent", you can later use that parent snapshot in combination with `zfs diff` to implement the storage-driver `Diff` interface method.

(i.e. below, the last line efficiently generates a list of added/changed/deleted files between layers):

```
zfs snapshot parent@whatever
zfs clone parent@whatever child
zfs snapshot child@parent
(import files, run commands, whatever)
zfs diff child@parent child
```

Mic92 commented:
@burke I have actually had implemented this, but unfortunately `zfs diff` only report changes once per inode.
 This breaks horrible if someone use hardlinks: 

```
RUN touch test && ln test test2
```
because `zfs diff` will have only report either test or test2 was added but not both.
The workaround would be to scan every changed directories, take the inode number of the files their and compare them with the inode reported by `zfs diff`. This could be implemented in the future. I don't know if it will faster then the current approach.

Mic92 commented:
@discordianfish so I will make a pull request once you have go through the whole documentation.

tiborvass commented:
@Mic92 #dontusehardlinks :P

burke commented:
Ohh, interesting. That's really unfortunate behaviour. Too bad.
